README.md
# Tarea1-Sistemas-Operativos

test/history.txt
   25  gcc test.c -o test && ./test 
   26  gcc test.c -o test && ./test 
   27  gcc test.c -o test && ./test 
   28  gcc test.c -o test && ./test 
   29  gcc test.c -o test && ./test 
   30  gcc test.c -o test && ./test 
   31  gcc test.c -o test && ./test 
   32  gcc test.c -o test && ./test 
   33  gcc test.c -o test && ./test 
   34  gcc test.c -o test 
   35  gcc test.c -o test && ./test 
   36  gcc test.c -o test && ./test 
   37  gcc test.c -o test && ./test 
   38  gcc test.c -o test && ./test 
   39  gcc test.c -o test && ./test 
   40  gcc test.c -o test
   41  ./test 
   42  ./test 
   43  gcc test.c -o test
   44  ./test 
   45  ./test 
   46  ./test 
   47  ./test 
   48  netstat -a -b
   49  ./test 
   50  netstat -a -b
   51  gcc test.c -o test
   52  gcc test.c -o test & ./test 
   53  gcc test.c -o test && ./test 
   54  gcc test.c -o test && ./test 
   55  gcc test.c -o test && ./test 
   56  gcc test.c -o test && ./test 
   57  gcc test.c -o test && ./test 
   58  gcc test.c -o test && ./test 
   59  gcc test.c -o test && ./test 
   60  gcc test.c -o test && ./test 
   61  gcc test.c -o test && ./test 
   62  gcc test.c -o test && ./test 
   63  gcc test.c -o test && ./test 
   64  gcc test.c -o test && ./test 
   65  gcc test.c -o test && ./test 
   66  gcc test.c -o test && ./test 
   67  gcc test.c -o test && ./test 
   68  gcc test.c -o test && ./test 
   69  gcc test.c -o test && ./test 
   70  gcc test.c -o test && ./test 
   71  gcc test.c -o test && ./test 
   72  gcc test.c -o test && ./test 
   73  gcc test.c -o test && ./test 
   74  gcc test.c -o test && ./test 
   75  gcc test.c -o test && ./test 
   76  gcc test.c -o test && ./test 
   77  gcc test.c -o test && ./test 
   78  gcc test.c -o test && ./test 
   79  gcc test.c -o test && ./test 
   80  gcc test.c -o test && ./test 
   81  gcc test.c -o test && ./test 
   82  gcc test.c -o test && ./test 
   83  gcc test.c -o test && ./test 
   84  gcc test.c -o test && ./test 
   85  gcc test.c -o test && ./test 
   86  gcc test.c -o test && ./test 
   87  cd t
   88  cd Tarea1-Sistemas-Operativos/.
   89  ls
   90  make 
   91  make 
   92  make 
   93  make 
   94  make 
   95  make 
   96  ls
   97  ./ImageServer 
   98  make clean
   99  ./ImageServer 
  100  make 
  101  make 
  102  ./ImageServer 
  103  make 
  104  make clean
  105  make 
  106  ./ImageServer 
  107  ./ImageServer 
  108  make clean
  109  make 
  110  make && ./ImageServer 
  111  make && ./ImageServer 
  112  make clean
  113  make && ./ImageServer 
  114  cat makefile 
  115  nano makefile 
  116  gcc ImageServer 
  117  make && ./ImageServer 
  118  gcc ImageServer 
  119  make clean
  120  make && ./ImageServer 
  121  make clean
  122  make && ./ImageServer 
  123  make clean
  124  make && ./ImageServer 
  125  make && ./ImageServer 
  126  make && ./ImageServer 
  127  make && ./ImageServer 
  128  make && ./ImageServer 
  129  make && ./ImageServer 
  130  make && ./ImageServer 
  131  make && ./ImageServer 
  132  make && ./ImageServer 
  133  make && ./ImageServer 
  134  make && ./ImageServer 
  135  make && ./ImageServer 
  136  make && ./ImageServer 
  137  make && ./ImageServer 
  138  make && ./ImageServer 
  139  make && ./ImageServer 
  140  make && ./ImageServer 
  141  make && ./ImageServer 
  142  make && ./ImageServer 
  143  make && ./ImageServer 
  144  make && ./ImageServer 
  145  make && ./ImageServer 
  146  make && ./ImageServer 
  147  make && ./ImageServer 
  148  make && ./ImageServer 
  149  make && ./ImageServer 
  150  make && ./ImageServer 
  151  make && ./ImageServer 
  152  make && ./ImageServer 
  153  make && ./ImageServer 
  154  make && ./ImageServer 
  155  make && ./ImageServer 
  156  make && ./ImageServer 
  157  make && ./ImageServer 
  158  make && ./ImageServer 
  159  make && ./ImageServer 
  160  make && ./ImageServer 
  161  make && ./ImageServer 
  162  make && ./ImageServer 
  163  make && ./ImageServer 
  164  make && ./ImageServer 
  165  make && ./ImageServer 
  166  make && ./ImageServer 
  167  make && ./ImageServer 
  168  make && ./ImageServer 
  169  make && ./ImageServer 
  170  make && ./ImageServer 
  171  make && ./ImageServer 
  172  /home/gsegura/Documents/TEC/2021/S1/operativos/tareas/1/Tarea1-Sistemas-Operativos/ImageServer 
  173  jupyter lab
  174   kdevelop! -s "{6ba96841-1f97-40be-9ebb-f1acc7dc1ec0}"
  175   kdevelop! -s "{6ba96841-1f97-40be-9ebb-f1acc7dc1ec0}"
  176  python
  177  yay jitsi
  178   kdevelop! -s "{6ba96841-1f97-40be-9ebb-f1acc7dc1ec0}"
  179   kdevelop! -s "{6ba96841-1f97-40be-9ebb-f1acc7dc1ec0}"
  180  cd /home/gsegura/Documents/TEC/2021/S1/operativos/tareas/1/Tarea1-Sistemas-Operativos
  181  yay kdevelop
  182  yay okteta
  183  make && ./ImageServer 
  184  which gdb
  185  make && ./ImageServer 
  186  make && ./ImageServer 
  187  ls
  188  cd ..
  189  ls
  190  curl -X POST -F 'image=@./rect16.png' http://localhost:1717
  191  curl -X POST -F 'image=@./rect16.png' http://localhost:1717
  192  curl -X POST -F 'image=@./test.c' http://localhost:1717
  193  curl -X POST -F 'image=@./test.c' http://localhost:1717
  194  curl -X POST -F 'image=@./test.c' http://localhost:1717
  195  curl -X POST -F 'image=@./test.c' http://localhost:1717
  196  curl -X POST -F 'image=@./test.c' http://localhost:1717
  197  curl -X POST -F 'image=@./test.c' http://localhost:1717
  198  curl -X POST -F 'image=@./test.c' http://localhost:1717
  199  curl -X POST -F 'image=@./test.c' http://localhost:1717
  200  curl -X POST -F 'image=@./test.c' http://localhost:1717

test.c
// Server side C program to demonstrate HTTP Server programming
#include <stdio.h>
#include <sys/socket.h>
#include <unistd.h>
#include <stdlib.h>
#include <netinet/in.h>
#include <string.h>

#define PORT 8085
int main(int argc, char const *argv[])
{
    int server_fd, new_socket; long valread;
    struct sockaddr_in address;
    int addrlen = sizeof(address);
    
    // Only this line has been changed. Everything is same.
    char *hello = "HTTP/1.1 200 OK\nContent-Type: text/plain\nContent-Length: 12\n\nHello world!";
    
    // Creating socket file descriptor
    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0)
    {
        perror("In socket");
        exit(EXIT_FAILURE);
    }
    

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY;
    address.sin_port = htons( PORT );
    
    memset(address.sin_zero, '\0', sizeof address.sin_zero);
    
    
    if (bind(server_fd, (struct sockaddr *)&address, sizeof(address))<0)
    {
        perror("In bind");
        exit(EXIT_FAILURE);
    }
    if (listen(server_fd, 10) < 0)
    {
        perror("In listen");
        exit(EXIT_FAILURE);
    }
    while(1)
    {
        printf("\n+++++++ Waiting for new connection ++++++++\n\n");
        if ((new_socket = accept(server_fd, (struct sockaddr *)&address, (socklen_t*)&addrlen))<0)
        {
            perror("In accept");
            exit(EXIT_FAILURE);
        }
        
        char buffer[3000000] = {0};
        valread = read( new_socket , buffer, 3000000);
        
        
        char * s0, *s1;
        s0 = strstr(buffer, "boundary=");      // search for string "hassasin" in buff
        int l0=strlen("boundary=");
        
        if (s0 != NULL)                     // if successful then s now points at "hassasin"
        {
            printf("Found string at index = %d\n", s0 - buffer);
            
            s1 = strstr(s0, "\n");      // search for string "hassasin" in buff

            char boundary[40];
            memset(boundary, '\0', sizeof(boundary));
            strncpy(boundary, s0+l0, s1-s0-l0);

            
//             printf("%s\n",s );
//             printf("%.*s\n",s1-s0-l0,s0+l0 );
            printf("BOUNDARY: %s\n",boundary );

        }                                  // index of "hassasin" in buff can be found by pointer subtraction
        else
        {
            printf("String not found\n");  // `strstr` returns NULL if search string not found
        }
            
        
        printf("________________________\n\n\n%s\n",buffer );
        write(new_socket , hello , strlen(hello));
        printf("------------------Hello message sent-------------------");
        close(new_socket);
    }
    return 0;
} 


test/server.c


// Server side C/C++ program to demonstrate Socket programming 
#include <unistd.h> 
#include <stdio.h> 
#include <sys/socket.h> 
#include <stdlib.h> 
#include <netinet/in.h> 
#include <string.h> 
#define PORT 8085
int main(int argc, char const *argv[]) 
{ 
    int server_fd, new_socket, valread; 
    struct sockaddr_in address; 
    int opt = 1; 
    int addrlen = sizeof(address); 
    char buffer[10240] = {0}; 
    char *hello = "Hello from server"; 
       
    // Creating socket file descriptor 
    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0) 
    { 
        perror("socket failed"); 
        exit(EXIT_FAILURE); 
    } 
       
    // Forcefully attaching socket to the port 8080 
    if (setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, 
                                                  &opt, sizeof(opt))) 
    { 
        perror("setsockopt"); 
        exit(EXIT_FAILURE); 
    } 
    address.sin_family = AF_INET; 
    address.sin_addr.s_addr = INADDR_ANY; 
    address.sin_port = htons( PORT ); 
       
    // Forcefully attaching socket to the port 8080 
    if (bind(server_fd, (struct sockaddr *)&address,  
                                 sizeof(address))<0) 
    { 
        perror("bind failed"); 
        exit(EXIT_FAILURE); 
    } 
    if (listen(server_fd, 3) < 0) 
    { 
        perror("listen"); 
        exit(EXIT_FAILURE); 
    } 
    while(1){
        if ((new_socket = accept(server_fd, (struct sockaddr *)&address,  
                        (socklen_t*)&addrlen))<0) 
        { 
            perror("accept"); 
            exit(EXIT_FAILURE); 
        } 
        valread = read( new_socket , buffer, 10240); 
        printf("1XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n\n%s\n",buffer ); 
        
        valread = read( new_socket , buffer, 10240); 
        printf("2XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n\n%s\n",buffer ); 
        
        valread = read( new_socket , buffer, 10240); 
        printf("3XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n\n%s\n",buffer ); 
        
        valread = read( new_socket , buffer, 10240); 
        printf("4XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n\n%s\n",buffer ); 
        
        valread = read( new_socket , buffer, 10240); 
        printf("5XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n\n%s\n",buffer ); 
        
        valread = read( new_socket , buffer, 10240); 
        printf("6XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n\n%s\n",buffer ); 
        
        
        send(new_socket , hello , strlen(hello) , 0 ); 
        printf("Hello message sent\n"); 
    }
    return 0; 
} 


uini.h
/*
uini - single header-file INI parser

Copyright (c) 2019 Ethan McTague

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

#pragma once
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/** Trim whitespace from start and end of string. */
static inline char *uini_trim (char *in) {
  if (!in) return NULL;
  while (isspace(*in)) in++;
  char *end = in + strlen(in) - 1;
  while (end > in && isspace(*end)) end--;
  end[1] = '\0';
  return in;
}

/** uini_parse - simple INI file parser
 *  For each property in the INI file `f`, `handler` is invoked, passing:
 *   - The section (no section = empty string) containing the property.
 *   - The name of the property.
 *   - The value of the property.
 *   - The user pointer passed as `user` to parseINI.
 *  Syntax errors are mostly ignored silently and flexibly.
 *  Failure to provide a value for a property will result in `value` being passed as NULL.
 */
static inline void uini_parse(FILE *f, void (*handler)(const char *section, const char *name, const char *value, void *user), void *user)
{
  char *section = strdup(""), *line = NULL, *clean = NULL, *context = NULL;
  size_t read = 0;

  while (getline(&line, &read, f) > 0) {
    char *clean = uini_trim(strtok_r(line, ";\r\n", &context));
    if (!clean || !*clean) continue;
    if (*clean == '[') {
      free(section);
      section = uini_trim(strdup(strtok_r(++clean, "]", &context)));
    } else {
      const char *key = uini_trim(strtok_r(clean, "=", &context));
      if (!*key) continue;
      handler(section, key, uini_trim(strtok_r(NULL, "\r\n", &context)), user);
    }
  }
  free(section);
  free(line);
}

config.h
#ifndef _TCONFIG_H_
#define _TCONFIG_H_
#include <stdbool.h>

typedef struct ini_entry_s
{
  char *key;
  char *value;
} ini_entry_s;

typedef struct ini_section_s
{
  char *name;
  ini_entry_s *entry;
  int size;
} ini_section_s;

typedef struct ini_table_s
{
  ini_section_s *section;
  int size;
} ini_table_s;

/**
 * @brief Creates an empty ini_table_s struct for writing new entries to.
 * @return ini_table_s*
 */
ini_table_s *ini_table_create();

/**
 * @brief Free up all the allocated resources in the ini_table_s struct.
 * @param table
 */
void ini_table_destroy(ini_table_s *table);

/**
 * @brief Creates an ini_table_s struct filled with data from the specified
 *        `file'.  Returns NULL if the file can not be read.
 * @param table
 * @param file
 * @return ini_table_s*
 */
bool ini_table_read_from_file(ini_table_s *table, const char *file);

/**
 * @brief Writes the specified ini_table_s struct to the specified `file'.
 *        Returns false if the file could not be opened for writing, otherwise
 *        true.
 * @param table
 * @param file
 * @return bool
 */
bool ini_table_write_to_file(ini_table_s *table, const char *file);

/**
 * @brief Creates a new entry in the `table' containing the `key' and `value'
 *        provided if it does not exist.  Otherwise, modifies an exsiting `key'
 *        with the new `value'
 * @param table
 * @param section_name
 * @param key
 * @param value
 */
void ini_table_create_entry(ini_table_s *table, const char *section_name,
                            const char *key, const char *value);

/**
 * @brief Checks for the existance of an entry in the specified `table'.  Returns
 *        false if the entry does not exist, otherwise true.
 * @param table
 * @param section_name
 * @param key
 * @return bool
 */
bool ini_table_check_entry(ini_table_s *table, const char *section_name,
                           const char *key);

/**
 * @brief Retrieves the unmodified value of the specified `key' in `section_name'.
 *        Returns NULL if the entry does not exist, otherwise a pointer to the 
 *        entry value data.
 * @param table
 * @param section_name
 * @param key
 * @return const char*
 */
const char *ini_table_get_entry(ini_table_s *table, const char *section_name,
                                const char *key);

/**
 * @brief Retrieves the value of the specified `key' in `section_name', converted
 *        to int.  Returns false on failure, otherwise true.
 * @param table
 * @param section_name
 * @param key
 * @param [out]value
 * @return int
 */
bool ini_table_get_entry_as_int(ini_table_s *table, const char *section_name,
                                const char *key, int *value);

/**
 * @brief Retrieves the value of the specified `key' in `section_name', converted
 *        to bool.  Returns false on failure, true otherwise.
 * @param table
 * @param section_name
 * @param key
 * @param [out]value
 * @return bool
 */
bool ini_table_get_entry_as_bool(ini_table_s *table, const char *section_name,
                                 const char *key, bool *value);

#endif //_TCONFIG_H_

main.c
// Tecnológico de Costa Rica
// Area Académica de Ingeniería en Computadores
// CE4303 Sistemas Operativos - Tarea 1
// Por Joseph Vargas y Gustavo Segura
#include <stdio.h>
#include <sys/socket.h>
#include <unistd.h>
#include <stdlib.h>
#include <netinet/in.h>
#include <string.h>
#include <signal.h>
#include <pthread.h>
#include <stdbool.h>
#include <dirent.h>
#include <sys/stat.h>
#include <libgen.h>
#include <time.h>
#include "uini.h"
#include "log.h"

#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"

#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "stb_image_write.h"


// Constants
#define CONFIG_FILE "/etc/server/server.conf"
#define PATH_MAX_STRING_SIZE 512
#define BUFFER_SIZE 1024




// Struct to contain the configuration
typedef struct
{
    int port;
    const char *save_dir;
    const char *colors_dir;
    const char *histo_dir;
    const char *log_file;
} ConfParams;

// Function definitions
parseHTTP(int* new_socket, char *buffer, const char *save_dir);
void *image_main(void *context);
void *server_main(void *context);
void RGB_to_grayscale(stbi_uc *image,int height, int width);
void equalization(stbi_uc *image,int height, int width);
int mkdir_p(const char *dir, const mode_t mode);

// Program running variable
bool running = true;
struct sigaction action;
// Handle stop
void handle_stop(int signum)
{
    log_info("Exiting from ImageServer...");
    running = false;
    exit(0);
}
// Handle the conf file
void conf_handler(const char *section, const char *key, const char *value, void *params_pointer)
{
    ConfParams *params = params_pointer;
    if (strcmp(key, "save_dir") == 0)
    {
        params->save_dir = strdup(value);
    }
    else if (strcmp(key, "colors_dir") == 0)
    {
        params->colors_dir = strdup(value);
    }
    else if (strcmp(key, "histo_dir") == 0)
    {
        params->histo_dir = strdup(value);
    }
    else if (strcmp(key, "log_file") == 0)
    {
        params->log_file = strdup(value);
    }
    else if (strcmp(key, "port") == 0)
    {
        params->port = atoi(value);
    }
}

int main(int argc, char const *argv[])
{
    signal(SIGTERM, handle_stop);

    ConfParams params;

    // Create conf object
    FILE *f = fopen(CONFIG_FILE, "r");
    if (!f)
    {
        log_fatal("Error: Config file not found");
        return EXIT_FAILURE;
    }

    uini_parse(f, conf_handler, &params);
    fclose(f);

    // Setup logging
    char log_dir_name[PATH_MAX_STRING_SIZE];
    strcpy(log_dir_name, params.log_file);
    mkdir_p(dirname(log_dir_name), 0755);
    FILE *log_file = fopen(params.log_file, "a");
    log_add_fp(log_file, 0);

    // Debug
    mkdir_p(params.save_dir, 0755);
    log_info("save_dir is: %s", params.save_dir);
    mkdir_p(params.colors_dir, 0755);
    log_info("colors_dir is: %s", params.colors_dir);
    mkdir_p(params.histo_dir, 0755);
    log_info("histo_dir is: %s", params.histo_dir);
    log_info("port is: %i", params.port);

    // Main threads
    pthread_t image_thread;
    pthread_t server_thread;
    pthread_create(&image_thread, NULL, image_main, &params);
    pthread_create(&server_thread, NULL, server_main, &params);

    pthread_join(image_thread, NULL);
    pthread_join(server_thread, NULL);

    return 0;
}

void *server_main(void *context)
{
    ConfParams *params = context;
    
    log_debug("PORT FROM SERVER THREAD: %i", params->port);
    log_debug("LOGFILE FROM SERVER THREAD: %s", params->log_file);
    int server_fd, new_socket;
    long valread;
    struct sockaddr_in address;
    int addrlen = sizeof(address);

    // Only this line has been changed. Everything is same.
    char *hello = "HTTP/1.1 200 OK\nContent-Type: text/plain\nContent-Length: 12\n\nHello world!\n";

    // Creating socket file descriptor
    if ((server_fd = socket(AF_INET, SOCK_STREAM, 0)) == 0)
    {
        perror("In socket");
        exit(EXIT_FAILURE);
    }

    address.sin_family = AF_INET;
    address.sin_addr.s_addr = INADDR_ANY;
    address.sin_port = htons(params->port);

    memset(address.sin_zero, '\0', sizeof address.sin_zero);

    if (bind(server_fd, (struct sockaddr *)&address, sizeof(address)) < 0)
    {
        perror("In bind");
        exit(EXIT_FAILURE);
    }
    if (listen(server_fd, 10) < 0)
    {
        perror("In listen");
        exit(EXIT_FAILURE);
    }
    while (running)
    {
        log_info("+++++++ Waiting for new connection ++++++++");
        if ((new_socket = accept(server_fd, (struct sockaddr *)&address, (socklen_t *)&addrlen)) < 0)
        {
            perror("In accept");
            exit(EXIT_FAILURE);
        }

        // char buffer[3000000] = {0};
        char *buffer = (char *)malloc(BUFFER_SIZE);
        memset(buffer, '\0', sizeof(char) * BUFFER_SIZE);

        
        
        parseHTTP(&new_socket,buffer,params->save_dir);

        // printf("________________________\n\n\n%s\n", buffer);
        free(buffer);
        write(new_socket, hello, strlen(hello));
        log_info("------------------Hello message sent-------------------");
        close(new_socket);
    }
}

int parseHTTP(int* new_socket, char* buffer, const char* save_dir)
{
    long valread;
    int content_length=0;
    valread = read(*new_socket, buffer, BUFFER_SIZE);
    
//     printf("____________BUFFER____________\n\n\n%s\n___________________________\n", buffer);
    
    char *s0, *s1;
    s0 = strstr(buffer, "x-www-form-urlencoded");
    int l0 = strlen("x-www-form-urlencoded");

    if (s0 != NULL)
    {
        char *pcontent = strstr(buffer, "Content-Length:");
        int ret = sscanf(pcontent, "Content-Length: %d\r\n", &content_length);
        log_debug("Content-Length: %i", content_length);

        char *file_start = strstr(buffer, "Content-Type:");
        
        if (file_start != NULL)
        {
            file_start = strstr(file_start + 1, "\n");

            
            FILE *fp;
            
            time_t seconds;
            seconds = time(NULL);
            
            
            char sbuf[PATH_MAX_STRING_SIZE];
            sprintf (sbuf, "%s/%ld.png", save_dir, seconds);
            
            
            fp = fopen(sbuf, "wb");
            log_info("Writing to: %s", sbuf);
            long read_bytes = 0;
            while(content_length>read_bytes){
                valread = read(*new_socket, buffer, BUFFER_SIZE);
                fwrite(buffer, sizeof(char), valread, fp);
                read_bytes=read_bytes+valread;
            }
            fclose(fp);
            if (read_bytes == 0)
            {
                log_warn("0 bytes read, file removed");
                remove(sbuf);
            }
        }
        else
        {
            log_error("file start error");
        }
        return 1;
    }
    else
    {
        return 0;
    }
}

void get_pixel(stbi_uc *image, size_t imageWidth, size_t x, size_t y, stbi_uc *r, stbi_uc *g, stbi_uc *b)
{
    *r = image[4 * (y * imageWidth + x) + 0];
    *g = image[4 * (y * imageWidth + x) + 1];
    *b = image[4 * (y * imageWidth + x) + 2];
}

void set_pixel(stbi_uc *image, size_t imageWidth, size_t x, size_t y, stbi_uc r, stbi_uc g, stbi_uc b)
{
    image[4 * (y * imageWidth + x) + 0] = r;
    image[4 * (y * imageWidth + x) + 1] = g;
    image[4 * (y * imageWidth + x) + 2] = b;
}

char *get_smallest_file(char *directory)
{
    DIR *d;
    struct dirent *dir;
    struct stat st;
    unsigned long int max_size = 0;
    char *current_file = malloc(sizeof(char) * PATH_MAX_STRING_SIZE);
    d = opendir(directory);
    if (d)
    {
        while ((dir = readdir(d)) != NULL)
        {
            if (dir->d_type == DT_DIR)
            {
                continue;
            }
            char path[PATH_MAX_STRING_SIZE];
            strcpy(path, directory);
            strcat(path, dir->d_name);
            stat(path, &st);

            if (max_size == 0 || st.st_size < max_size)
            {
                max_size = st.st_size;
                strcpy(current_file, path);
            }
        }
        closedir(d);
    }
    if (max_size == 0)
        return "\0";
    return current_file;
}

char *concat_path(char *color, char *basepath)
{
    char *output = malloc(sizeof(char) * PATH_MAX_STRING_SIZE);
    strcpy(output, basepath);
    strcat(output, color);
    return output;
}

char *output_path_builder(char *filename, char *color, char *basepath)
{
    char *output = malloc(sizeof(char) * PATH_MAX_STRING_SIZE);
    strcpy(output, basepath);
    strcat(output, color);
    strcat(output, "/");
    strcat(output, filename);
    strcat(output, ".png");
    return output;
}

char *histo_output_path_builder(char *filename, char *basepath)
{
    char *output = malloc(sizeof(char) * PATH_MAX_STRING_SIZE);
    strcpy(output, basepath);
    strcat(output, filename);
    strcat(output, ".png");
    return output;
}

/* recursive mkdir */
/* FROM: https://gist.github.com/ChisholmKyle/0cbedcd3e64132243a39 */
int mkdir_p(const char *dir, const mode_t mode)
{
    char tmp[PATH_MAX_STRING_SIZE];
    char *p = NULL;
    struct stat sb;
    size_t len;

    /* copy path */
    len = strnlen(dir, PATH_MAX_STRING_SIZE);
    if (len == 0 || len == PATH_MAX_STRING_SIZE)
    {
        return -1;
    }
    memcpy(tmp, dir, len);
    tmp[len] = '\0';

    /* remove trailing slash */
    if (tmp[len - 1] == '/')
    {
        tmp[len - 1] = '\0';
    }

    /* check if path exists and is a directory */
    if (stat(tmp, &sb) == 0)
    {
        if (S_ISDIR(sb.st_mode))
        {
            return 0;
        }
    }

    /* recursive mkdir */
    for (p = tmp + 1; *p; p++)
    {
        if (*p == '/')
        {
            *p = 0;
            /* test path */
            if (stat(tmp, &sb) != 0)
            {
                /* path does not exist - create directory */
                if (mkdir(tmp, mode) < 0)
                {
                    return -1;
                }
            }
            else if (!S_ISDIR(sb.st_mode))
            {
                /* not a directory */
                return -1;
            }
            *p = '/';
        }
    }
    /* test path */
    if (stat(tmp, &sb) != 0)
    {
        /* path does not exist - create directory */
        if (mkdir(tmp, mode) < 0)
        {
            return -1;
        }
    }
    else if (!S_ISDIR(sb.st_mode))
    {
        /* not a directory */
        return -1;
    }
    return 0;
}

char *filename_without_extension(char *file_name)
{
    char *remaining_str;
    char *last_ext;
    if (file_name == NULL)
        return NULL;
    if ((remaining_str = malloc(strlen(file_name) + 1)) == NULL)
        return NULL;
    strcpy(remaining_str, basename(file_name));
    last_ext = strrchr(remaining_str, '.');
    if (last_ext != NULL)
        *last_ext = '\0';
    return remaining_str;
}

void *image_main(void *context)
{
    ConfParams *params = context;

    log_debug("PORT FROM IMG THREAD: %i", params->port);
    log_debug("LOGFILE FROM IMG THREAD: %s", params->log_file);

    // Create folders if they don't exist
    mkdir_p(concat_path("r", params->colors_dir), 0755);
    mkdir_p(concat_path("g", params->colors_dir), 0755);
    mkdir_p(concat_path("b", params->colors_dir), 0755);

    while (running)
    {
        char *smallest_file = get_smallest_file(params->save_dir);
        if (smallest_file[0] == '\0')
        {
            log_info("No file found in save_dir, skipping");
            sleep(5);
            continue;
        }
        log_info("Smallest file: %s\n", get_smallest_file(params->save_dir));
        int width, height; // image width, heigth,
        stbi_uc *image = stbi_load(get_smallest_file(params->save_dir), &width, &height, NULL, 4);
        log_debug("w: %i", width);
        log_debug("h: %i", height);

        stbi_uc r, g, b;
        unsigned long int r_sum = 0, g_sum = 0, b_sum = 0;

        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                get_pixel(image, width, x, y, &r, &g, &b);
                r_sum += r;
                g_sum += g;
                b_sum += b;
            }
        }
        log_debug("r: %li", r_sum);
        log_debug("g: %li", g_sum);
        log_debug("b: %li", b_sum);
        log_debug("output colors: %s", output_path_builder(filename_without_extension(smallest_file), "b", params->colors_dir));

        // Case Red
        if (r_sum > g_sum && r_sum > b_sum)
        {
            log_info("RED!!");
            stbi_write_png(output_path_builder(filename_without_extension(smallest_file), "r", params->colors_dir), width, height, 4, image, 4 * width);
        }
        // Case Green
        else if (g_sum > r_sum && g_sum > b_sum)
        {
            log_info("GREEN!!");
            stbi_write_png(output_path_builder(filename_without_extension(smallest_file), "g", params->colors_dir), width, height, 4, image, 4 * width);
        }
        // Case Blue
        else
        {
            log_info("BLUE!!");
            stbi_write_png(output_path_builder(filename_without_extension(smallest_file), "b", params->colors_dir), width, height, 4, image, 4 * width);
        }

        equalization(image, height, width);
        log_debug("Equalization done");
        log_debug("output eq: %s", histo_output_path_builder(filename_without_extension(smallest_file), params->histo_dir));
        stbi_write_png(histo_output_path_builder(filename_without_extension(smallest_file), params->histo_dir), width, height, 4, image, 4 * width);

        remove(smallest_file);
        sleep(1);
    }
}

// https://stackoverflow.com/questions/15686277/convert-rgb-to-grayscale-in-c#15686412
double sRGB_to_linear(double x)
{
    if (x < 0.04045)
        return x / 12.92;
    return pow((x + 0.055) / 1.055, 2.4);
}

void RGB_to_grayscale(stbi_uc *image,int height, int width)
{
    stbi_uc r, g, b;
    for (int x = 0; x < width; x++)
    {
        for (int y = 0; y < height; y++)
        {
           get_pixel(image, width, x, y, &r, &g, &b);
            double R_linear = sRGB_to_linear(r / 255.0);
            double G_linear = sRGB_to_linear(g / 255.0);
            double B_linear = sRGB_to_linear(b / 255.0);
            double gray_linear = 0.2126 * R_linear + 0.7152 * G_linear + 0.0722 * B_linear;
            
             r = round(gray_linear*255);
             g = round(gray_linear*255);
             b = round(gray_linear*255);
            
            set_pixel(image, width,  x,  y,  r,  g,  b);
        }
    }
}

void equalization(stbi_uc *image,int height, int width){
    int fi[256]={0},cuf[256]={0}, cufeq[256]={0}, i_out[256]={0};
    stbi_uc r, g, b;
    RGB_to_grayscale(image, height,  width);
    
    for (int x = 0; x < width; x++)
    {
        for (int y = 0; y < height; y++)
        {
            get_pixel(image, width, x, y, &r, &g, &b);
            fi[r]=fi[r]+1;
//             set_pixel(image, width,  x,  y,  r,  g,  b);
        }
    }
    
    int sum=0;
    int i;
   for( i = 0; i < 256; i = i + 1 ){
       sum= sum+ fi[i];
       cuf[i]=sum;
   }
   int mean = sum>>8;
   int cuf_sum=0;
   if((sum&255)!=0){
       cuf_sum=1;
    }
    for( i = 0; i < 256; i = i + 1 ){
       cuf_sum= mean + cuf_sum;
       cufeq[i]=cuf_sum;
   }
   
   for( i = 0; i < 256; i = i + 1 ){
        int closest=0;
        if(cuf[i]<=cufeq[0]){
           i_out[i]=0;
           continue;
        }
        int j;
        for( j = 1; j < 256; j = j + 1 ){
            if(cuf[i]==cufeq[j]){
                i_out[i]=j; 
                break;
            }
            if(cuf[i]<cufeq[j]){
                if( (cuf[i]-cufeq[j])  >  (cuf[i]<cufeq[j-1])    ){
                    i_out[i]=j-1; 
                }
                else{
                    i_out[i]=j;
                }
                break;
            }
            if(j==255){
                i_out[i]=255;
            }
        }
   }
   for (int x = 0; x < width; x++)
    {
        for (int y = 0; y < height; y++)
        {
            get_pixel(image, width, x, y, &r, &g, &b);
            set_pixel(image, width,  x,  y,  i_out[r],   i_out[r],   i_out[r]);
        }
    }
}
